phoenix-core/src/main/java/org/apache/phoenix/coprocessor/StoreCompactionScanner.java

```diff
@@ -0,0 +1,422 @@
+                        return false;
+                    }
+                    rowVersion = new CompactionRowVersion();
+                    rowVersions.add(rowVersion);
+                    rowVersion.ts = ts;
+                    if (lastDeleteFamilyVersionMarker != null) {
+                        rowVersion.deleteFamilyMarker = lastDeleteFamilyVersionMarker;
+                        // The delete family version delete marker is consumed, so set it to null
+                        lastDeleteFamilyVersionMarker = null;
+                    } else if (lastDeleteFamilyMarker != null) {
+                        rowVersion.deleteFamilyMarker = lastDeleteFamilyMarker;
+                    }
+                    if (columnDeleteMarkers != null) {
+                        rowVersion.columnDeleteMarkers = new ArrayList<>(columnDeleteMarkers);
+                    }
+                    rowVersion.cells.add(firstCell);
+                    rowVersion.version = version++;
+                } else if (firstCell.getTimestamp() == ts) {
+                    rowVersion.cells.add(firstCell);
+                } else if (columnDeleteMarkers == null ||
+                        !isCellDeleted(columnDeleteMarkers, rowVersion, firstCell, storeLevel)) {
+                    rowVersion.cells.add(firstCell);
+                }
+            }
+            // Remove the columns that are empty
+            int columnIndex = 0;
+            while (!columns.isEmpty() && columnIndex < columns.size()) {
+                if (columns.get(columnIndex).isEmpty()) {
+                    columns.remove(columnIndex);
+                } else {
+                    columnIndex++;
+                }
+            }
+        }
+        return true;
+    }
+
+    private void addCells(CompactionRowVersion rowVersion, List<Cell> result) {
+        if (rowVersion.columnDeleteMarkers == null) {
+            result.addAll(rowVersion.cells);
+            return;
+        }
+        for (Cell cell : rowVersion.cells) {
+            if (rowVersion.columnDeleteMarkers.isEmpty()) {
+                result.add(cell);
+            }
+            else {
+                Cell dm = rowVersion.columnDeleteMarkers.get(0);
+                if (Bytes.compareTo(cell.getQualifierArray(), cell.getQualifierOffset(),
+                        cell.getQualifierLength(), dm.getQualifierArray(), dm.getQualifierOffset(),
+                        dm.getQualifierLength()) == 0) {
+                    rowVersion.columnDeleteMarkers.remove(0);
+                    continue;
+                }
+                else {
+                    result.add(cell);
+                }
+            }
+        }
+    }
+    private void updateResultsWithRowOutsideMaxLookbackButInsideTTLWindow(List<Cell> result,
+            CompactionRowVersion rowVersion) {
+        // The rows that are outside the max lookback window but inside the TTL window
+        if (rowVersion.deleteFamilyMarker == null) {
```

Comment:
Do we need any logic about minVersions in here, or does the fact that minVersions < maxVersions mean that the rowVersion.version < maxVersion check also takes care of minVersion?

Reply:
The MinVersions check is required to be done outside the TTL window. If the version outside the TTL window is less than MinVersions then it is retained.